Fast refresh indices should use search shards

kingherc opened the pull request:
Fast refresh indices should now behave like non fast refresh indices in how they execute  (m)gets and searches. I.e., they should use the search shards.

For BWC, we define a new transport version. We expect search shards to be upgraded first, before promotable shards. Until the cluster is fully upgraded, the promotable shards (whether upgraded or not) will still receive and execute gets/searches locally.

Relates ES-9573
Relates ES-9579

elasticsearchmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

arteam reviewed (approved):
LGTM! Looking forward to simplify things with ES-9563 after this PR is successfully rolled out.

henningandersen commented on server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportUnpromotableShardRefreshAction.java:
```diff
@@ -73,6 +76,17 @@ protected void unpromotableShardOperation(
 
+        // During an upgrade to FAST_REFRESH_RCO, we expect search shards to be first upgraded before the primary is upgraded. Thus,
+        // when the primary is upgraded, and starts to deliver unpromotable refreshes, we expect the search shards to be upgraded already.
+        // TODO: remove assertion (ES-9563)
```
I'd note that the fast refresh setting is final, since otherwise this could in theory fail.

henningandersen commented on server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java:
```diff
@@ -125,13 +126,24 @@ protected void asyncShardOperation(GetRequest request, ShardId shardId, ActionLi
+            // TODO: remove assertion (ES-9563)
             assert indexShard.indexSettings().isFastRefresh() == false
-                : "a search shard should not receive a TransportGetAction for an index with fast refresh";
+                || clusterService.state().getMinTransportVersion().onOrAfter(TransportVersions.FAST_REFRESH_RCO)
```
I am not sure this assertion is safe. We risk another node seeing seeing that the min-transport-version advanced prior to this search node realizing it. And a GET request could thus arrive here ahead of the node knowing about the new  min-transport version.

henningandersen commented on server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java:
```diff
@@ -125,13 +126,24 @@ protected void asyncShardOperation(GetRequest request, ShardId shardId, ActionLi
+        // TODO: adapt assertion to assert only that it is not stateless (ES-9563)
+        assert DiscoveryNode.isStateless(clusterService.getSettings()) == false
+            || (indexShard.indexSettings().isFastRefresh()
+                && clusterService.state().getMinTransportVersion().before(TransportVersions.FAST_REFRESH_RCO))
```
Similarly, here, a GET message for a fast refresh shard could be sent from a node where the min-transport-version had not advanced but then received on this search node after it advances, making this assertion unsafe.

henningandersen commented on server/src/main/java/org/elasticsearch/action/support/replication/PostWriteRefresh.java:
```diff
@@ -53,9 +52,7 @@ public void refreshShard(
-                    // Fast refresh indices do not depend on the unpromotables being refreshed
-                    boolean fastRefresh = IndexSettings.INDEX_FAST_REFRESH_SETTING.get(indexShard.indexSettings().getSettings());
-                    if (location != null && (indexShard.routingEntry().isSearchable() == false && fastRefresh == false)) {
+                    if (location != null && indexShard.routingEntry().isSearchable() == false) {
```
This fixes it for future refreshes after the indexing node upgraded. But it does not guarantee immediate availability of the latest state on the search node. So we risk some seconds of non-realtime GET requests going backwards during such an upgrade? I think real-time GET requests will be saved by the wait-for generation, is that also your understanding?

kingherc replied:
The reasoning here is this code runs on the primary/indexing node, and indeed that the indexing node will be upgraded after the search nodes.

> But it does not guarantee immediate availability of the latest state on the search node.

Doesn't our upgrade process guarantee that, since search nodes are upgraded first?

> So we risk some seconds of non-realtime GET requests going backwards during such an upgrade?

A non-realtime GET coordinated by an old search node will go the primary to execute.
A non-realtime GET coordinated by a new search node, with an old primary node, will go the primary to execute.
A non-realtime GET coordinated by a new search node on a fully upgraded cluster, will be executed on the search node as is done for non-fast-refresh indices. Which should be fine as well. Not sure I see when/why it might go backwards?

> I think real-time GET requests will be saved by the wait-for generation, is that also your understanding?

A real-time GET coordinated by an old search node will go the primary to execute.
A real-time GET coordinated by a new search node, with an old primary node, will go the primary to execute.
A real-time GET coordinated by a new search node on a fully upgraded cluster, will be executed on the search node as is done for non-fast-refresh indices. Which should use wait-for generation if necessary.

Please tell me if you see any corner cases I might have missed or not considered. It might be useful to think about the above combinations also for searches/mgets, but I believe it should be a similar story for them as well.

henningandersen replied:
I think you are right that it works out. The upgrade will force a relocation, which forces a flush, bringing things back into order. Thanks.

henningandersen commented on server/src/main/java/org/elasticsearch/index/cache/bitset/BitsetFilterCache.java:
```diff
@@ -105,7 +105,7 @@ static boolean shouldLoadRandomAccessFiltersEagerly(IndexSettings settings) {
         boolean isStateless = DiscoveryNode.isStateless(settings.getNodeSettings());
         if (isStateless) {
-            return DiscoveryNode.hasRole(settings.getNodeSettings(), DiscoveryNodeRole.INDEX_ROLE)
+            return DiscoveryNode.hasRole(settings.getNodeSettings(), DiscoveryNodeRole.SEARCH_ROLE)
                 && loadFiltersEagerlySetting
                 && INDEX_FAST_REFRESH_SETTING.get(settings.getSettings());
```
This seems odd at first sight, as if we still rely on the fast refresh setting?

henningandersen replied:
I suppose the change here is not *that bad* to prevent merging this (seems similarly wrong, assuming my understanding is correct).

kingherc replied:
I tried to keep the behavior of the `if` as it's in the code currently, but I agree that PR's final modifications seem inconsistent with its description, so if it's a bug it should be corrected. Preferably in an orthogonal PR and I rebase afterwards. But lets see what the Search team comments first.

original-brownbear replied:
Yes the idea here was to only eager load when something gets searched. If we know we will never search on an indexing node then eager loading is unnecessary there :)

kingherc replied:
Thanks @original-brownbear ! As it's orthogonal to this PR, and to avoid any confusion, I believe it's better to fix this in a separate PR. I opened https://github.com/elastic/elasticsearch/pull/114191 for our convenience, please feel free to review -- or make a brand new one if you'd like.

After that PR is merged, I will rebase this one.

henningandersen replied:
I'd be ok to merge this PR with this change as is - it seems it does not make it worse (and the other PR can then resolve it).

kingherc replied:
Thanks @henningandersen ! I can merge after the stateless PR is also approved. If this is merged first, indeed I can rework/simplify https://github.com/elastic/elasticsearch/pull/114191 afterwards.

kingherc replied:
Fixed in https://github.com/elastic/elasticsearch/pull/114191 .

henningandersen commented on server/src/main/java/org/elasticsearch/action/get/TransportShardMultiGetAction.java:
```diff
@@ -124,14 +125,24 @@ protected void asyncShardOperation(MultiGetShardRequest request, ShardId shardId
-                : "a search shard should not receive a TransportShardMultiGetAction for an index with fast refresh";
+                || clusterService.state().getMinTransportVersion().onOrAfter(TransportVersions.FAST_REFRESH_RCO)
+                : "a search shard in a cluster with min transport version "
+                    + clusterService.state().getMinTransportVersion()
```
Similar assertion worries as above.

henningandersen reviewed (commented):
I have a few comments/questions.

kingherc reviewed (commented):
Thanks @henningandersen for the feedback! Feel free to review again.

kingherc committed:
Fast refresh indices should use search shards

Fast refresh indices should now behave like non fast refresh indices in
how they execute (m)gets and searches. I.e., they should use the search
shards.

For BWC, we define a new transport version. We expect search shards to
be upgraded first, before promotable shards. Until the cluster is fully
upgraded, the promotable shards (whether upgraded or not) will still
receive and execute gets/searches locally.

Relates ES-9573
Relates ES-9579

henningandersen reviewed (commented):
Looks good, main issue remaining is the BitsetFilterCache.

kingherc commented on server/src/main/java/org/elasticsearch/action/get/TransportShardMultiGetAction.java:
```diff
@@ -124,14 +124,13 @@ protected void asyncShardOperation(MultiGetShardRequest request, ShardId shardId
         assert DiscoveryNode.isStateless(clusterService.getSettings()) == false || indexShard.indexSettings().isFastRefresh()
-            : "in Stateless a promotable to primary shard can receive a TransportShardMultiGetAction only if an index has "
-                + "the fast refresh setting";
+            : "in Stateless a promotable to primary shard can receive a TransportShardMultiGetAction only if an index has the fast "
```
The change just splits the string. Probably some effect of spotless. I wouldn't consider it important, right?

kingherc replied:
Reverted

kingherc committed:
Revert comment

kingherc committed:
Merge remote-tracking branch 'kingherc/main' into non-issue/ES-9573-fast-refresh-rco

henningandersen reviewed (approved):
LGTM.

kingherc committed:
Merge remote-tracking branch 'kingherc/main' into non-issue/ES-9573-fast-refresh-rco

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.x](https://github.com/elastic/elasticsearch/pull/114259)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/114259">](https://github.com/elastic/elasticsearch/pull/114259) |

ywangd commented:
IIUC, it is not absolutely necessary to backport this PR to 8.16 since the change affects serverless only and serverless works on the `main` branch only? I am trying understanding the reason here in case it applies to any future work. Thanks!

ywangd commented:
Thanks for the explanation. 🙏
